src/ucx_plugin.c

```diff
@@ -276,114 +278,123 @@ static ncclResult_t ucx_worker_get_netaddress(ucp_worker_h worker,
 }
 
 static ncclResult_t ucx_get_ctx_and_worker(int dev, ucp_context_h *ctx,
-                                           ucp_worker_h *worker,
+                                           nccl_ucx_worker_t **ucx_worker,
                                            ucp_tag_t *newtag) {
   pthread_mutex_lock(&nccl_ucx_lock);
-  if (ncclNIbDevs < dev) {
+  if (ncclNIbDevs <= dev) {
     WARN("Device index is too large");
-    return ncclSystemError;
+    goto err;
   }
 
-  if (workers[dev].count == 0) {
-    ucx_init_context(&workers[dev].ctx, dev);
-    ucx_init_worker(workers[dev].ctx, &workers[dev].worker);
-    workers[dev].last_tag = tag;
-    worker_count++;
+  nccl_ucx_worker_t *w;
+  for (w = workers[dev]; w != NULL; w = w->next) {
+    if (w->dev == dev && w->thread == pthread_self()) {
```

Comment:
is it possible that `w->dev != dev` here?

Reply:
makes sense, moved as assert